### PR TITLE
Closes #438.

### LINKED ISSUE #438: Proposal to add MedNLI dataset
## Adding a Dataset
- **Name:** MedNLI - A Natural Language Inference Dataset For The Clinical Domain
- **Description:** A dataset annotated by doctors, performing a natural language inference task), grounded in the medical history of patients.
- **Task:** Natural Language Inference
- **Paper:** https://arxiv.org/abs/1808.06752
- **Data:** https://physionet.org/content/mednli/1.0.0/
- **License:** PhysioNet license
- **Motivation:** A good dataset annotated by clinicians with thousands of sentence pairs.

### PR DESCRIPTION
**Pull request by JohnGiorgi**
- **Name:** MedNLI
- **Description:** MedNLI - a dataset annotated by doctors, performing a natural language inference task (NLI),
grounded in the medical history of patients.
- **Paper:** https://physionet.org/content/mednli/1.0.0/
- **Data:** https://physionet.org/content/mednli/1.0.0/

### Checkbox

- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `biodatasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_BIGBIO_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `BigBioConfig` for the source schema and one for a bigbio schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_bigbio biodatasets/my_dataset/my_dataset.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

Closes #438.

### CODE REVIEW

**Comment by JohnGiorgi** [comment]
Here is the output of the unit tests:

```
python -m tests.test_bigbio biodatasets/mednli/mednli.py --data_dir /Users/johngiorgi/Documents/dev/biomedical/physionet.org/files/mednli/1.0.0
INFO:__main__:args: Namespace(data_dir='/Users/johngiorgi/Documents/dev/biomedical/physionet.org/files/mednli/1.0.0', path='biodatasets/mednli/mednli.py', schema=None, subset_id=None, use_auth_token=None)
INFO:__main__:self.PATH: biodatasets/mednli/mednli.py
INFO:__main__:self.SUBSET_ID: mednli
INFO:__main__:self.SCHEMA: None
INFO:__main__:self.DATA_DIR: /Users/johngiorgi/Documents/dev/biomedical/physionet.org/files/mednli/1.0.0
INFO:__main__:Checking for _SUPPORTED_TASKS ...
INFO:__main__:Found _SUPPORTED_TASKS=[<Tasks.TEXTUAL_ENTAILMENT: 'TE'>]
INFO:__main__:_SUPPORTED_TASKS implies _MAPPED_SCHEMAS={'TE'}
INFO:__main__:schemas_to_check: {'TE'}
INFO:__main__:Checking load_dataset with config name mednli_source
WARNING:datasets.builder:Using custom data configuration mednli_source-e74e1630bc2bbeba
WARNING:datasets.builder:Reusing dataset med_nli_dataset (/Users/johngiorgi/.cache/huggingface/datasets/med_nli_dataset/mednli_source-e74e1630bc2bbeba/1.0.0/1caf5757196c0609c9086428cd2bfb1e7ee9c4534b8e10cdcfb2034a7766d5f5)
100%|█████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 375.99it/s]
INFO:__main__:Checking load_dataset with config name mednli_bigbio_te
WARNING:datasets.builder:Using custom data configuration mednli_bigbio_te-e74e1630bc2bbeba
WARNING:datasets.builder:Reusing dataset med_nli_dataset (/Users/johngiorgi/.cache/huggingface/datasets/med_nli_dataset/mednli_bigbio_te-e74e1630bc2bbeba/1.0.0/1caf5757196c0609c9086428cd2bfb1e7ee9c4534b8e10cdcfb2034a7766d5f5)
100%|█████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, 917.19it/s]
INFO:__main__:Checking global ID uniqueness
INFO:__main__:Found 1395 unique IDs
INFO:__main__:Gathering schema statistics
INFO:__main__:Gathering schema statistics
train
==========
id: 11232
premise: 11232
hypothesis: 11232
label: 11232

test
==========
id: 1422
premise: 1422
hypothesis: 1422
label: 1422

validation
==========
id: 1395
premise: 1395
hypothesis: 1395
label: 1395

.
----------------------------------------------------------------------
Ran 1 test in 3.125s

OK
```

File: biodatasets/mednli/mednli.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        if self.config.data_dir is None:
+            raise ValueError("This is a local dataset. Please pass the data_dir kwarg to load_dataset.")
+        else:
+            data_dir = self.config.data_dir
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, "mli_train_v1.jsonl"),
```
Comment (galtay) [inline biodatasets/mednli/mednli.py:140]: in the other local dataset loaders we've tried to use the tar/zip/archive files directly. would it be possible for this script to work directly with the `mednli-a-natural-language-inference-dataset-for-the-clinical-domain-1.0.0.zip` file?

File: biodatasets/mednli/mednli.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+mapping from inputs to outputs. However, they still lack generalization capabilities in conditions
+that differ from the ones encountered during training. This is even more challenging in specialized,
+and knowledge intensive domains, where training data is limited. To address this gap, we introduce
+MedNLI - a dataset annotated by doctors, performing a natural language inference task (NLI),
+grounded in the medical history of patients. As the source of premise sentences, we used the
+MIMIC-III. More specifically, to minimize the risks to patient privacy, we worked with clinical
+notes corresponding to the deceased patients. The clinicians in our team suggested the Past Medical
+History to be the most informative section of a clinical note, from which useful inferences can be
+drawn about the patient.
+"""
```
Comment (galtay) [inline biodatasets/mednli/mednli.py:27]: please add some information to the docstring concerning the form that the input should take (e.g. https://github.com/bigscience-workshop/biomedical/blob/master/biodatasets/n2c2_2006_deid/n2c2_2006_deid.py#L34)

**Review by galtay — CHANGES_REQUESTED** [review]

**Review by galtay — CHANGES_REQUESTED** [review]
a few changes about the data_dir and file descriptions please

**Comment by JohnGiorgi** [comment]
@galtay This has all the same fixes as #468 so I think its ready to merge unless you spot anything?

**Review by galtay — APPROVED** [review]
thank you for working through my confusing requests! LGTM

**Comment by JohnGiorgi** [comment]
> thank you for working through my confusing requests! LGTM

Thanks for all the feedback!!

**Comment by NtaylorOX** [comment]
Do you have an example of using this using the load_dataset function? As when I try it fails as seems to not like the data_dir being passed as an argument, 

e.g. 

```
from datasets import load_dataset

data_dir="/mnt/sdd/niallt/mednli-a-natural-language-inference-dataset-for-the-clinical-domain-1.0.0.zip"
dataset = load_dataset("bigbio/mednli", data_dir = data_dir)

```
leads to:

AttributeError: 'BuilderConfig' object has no attribute 'schema'

And the documentation of how to actually provide the data_dir argument is not all that clear from what I can tell.

**Pull request merged**